src/libleo/api/getkadr.s

```diff
@@ -0,0 +1,229 @@
+    li      $at, 0x8800
+    slt     $at, $a0, $at
+    bnez    $at, .nonkanji
+     // 0xBC is number of glyphs in one block in the `kanji` file:
+     // 0x100 possible codepoints with the same byte1
+     // - 0x40 unused at beginning
+     // - 1 unused at 0x7F
+     // - 3 unused at 0xFD, 0xFE, 0xFF
+     li     $a2, 0xBC
+    // Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x88
+    multu   $a2, $a1
+    // Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xFF
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, (0x80 - 0x40)
+    mflo    $a2
+    // 0x__7F is always empty and elided in the file, so decrement if larger
+    bnezl   $at, .kanji_lower_halfblock
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.kanji_lower_halfblock:
+    addi    $a3, $a3, 0x30A
+    add     $a3, $a3, $a2
+    jr      $ra
+     sll    $v0, $a3, 7
+     // returns (0x30A + (adjusted byte2) + (adjusted byte1) * 0xBC) * FONT_CHAR_TEX_SIZE
+
+// Non-kanji are arranged with irregular gaps, use the lookup table.
+.nonkanji:
+    // Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x81
+    multu   $a2, $a1
+    // Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xFF
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, (0x80 - 0x40)
+    mflo    $a2
+    // 0x__7F is always empty and elided in the file, so decrement if larger
+    bnezl   $at, .nonkanji_lower_halfblock
+     mflo   $a2
+    addi    $a3, $a3, -1
+    mflo    $a2
+.nonkanji_lower_halfblock:
+    add     $a3, $a3, $a2
+    lui     $a2, %hi(sLeoNonKanjiIndices)
+    sll     $a3, $a3, 1
+    addiu   $a2, %lo(sLeoNonKanjiIndices)
+    add     $a3, $a3, $a2
+    lh      $a2, ($a3)
+    jr      $ra
+     sll    $v0, $a2, 7
+     // returns sLeoNonKanjiIndices[(adjusted byte2) + (adjusted byte1) * 0xBC] * FONT_CHAR_TEX_SIZE
+.out_of_range:
+    jr      $ra
+     li     $v0, -1
+END(LeoGetKAdr)
+
+/**
+ * Nearly equivalent C code (Equivalent for all valid input, will behave differently on overflow due to the use of `add` in the original):
+ *
+ * extern u16 sLeoNonKanjiIndices[];
+ *
+ * s32 LeoGetKAdr(s32 sjis) {
+ *     u32 byte1;
+ *     u32 byte2;
+ *
+ *     if (sjis < 0x8140 || sjis > 0x9872) {
+ *         return -1;
+ *     }
+ *
+ *     byte1 = (u32)sjis >> 8;
+ *     byte2 = sjis & 0xFF;
+ *
+ *     byte2 -= 0x40;
+ *     if (byte2 >= 0x40) {
+ *         byte2--;
+ *     }
+ *
+ *     if (sjis >= 0x8800) {
+ *         byte1 -= 0x88;
+ *         return (0x30A + byte2 + byte1 * 0xBC) * FONT_CHAR_TEX_SIZE;
+ *     } else {
+ *         byte1 -= 0x81;
+ *         return sLeoNonKanjiIndices[byte2 + byte1 * 0xBC] * FONT_CHAR_TEX_SIZE;
+ *     }
+ * }
+ */
+
+/**
+ * Index table, formatting follows conventional Shift-JIS character table layout,
+ * with space for the missing 0x__7F,0x__FD,0x__FE,0x__FF entries, and split
+ * into blocks by high byte.
+ */
+DATA(sLeoNonKanjiIndices)
```

Comment:
The original name is `kantbl`. I'd prefer we use it and maybe make this label static, but if you're strongly against using it I think it at least deserves a comment.

Reply:
Thanks. Should we rename the labels to `kanchg1` etc too?

Reply:
I'm not so bothered about the label names, maybe at most we can comment the originals but it doesn't have to happen now.